src/role/index.js

```diff
@@ -0,0 +1,68 @@
+import { EventEmitter } from 'events';
+import shortId from 'shortid';
+import PHASE from './phase';
+import { assertType, is } from '../errors/runtime/type-assertions';
+import wrapTestFunction from '../api/wrap-test-function';
+import ensureUrlProtocol from '../utils/ensure-url-protocol';
+import { NavigateToCommand } from '../test-run/commands/actions';
+import roleMarker from './marker-symbol';
+import delay from '../utils/delay';
+
+const COOKIE_SYNC_DELAY = 100;
+
+class Role extends EventEmitter {
+    constructor (loginPage, initFn) {
+        super();
+
+        this[roleMarker] = true;
+
+        this.id    = shortId.generate();
+        this.phase = loginPage ? PHASE.uninitialized : PHASE.initialized;
+
+        this.loginPage = loginPage;
+        this.initFn    = initFn;
+
+        this.stateSnapshot = null;
+        this.initErr       = null;
+    }
+
+    async initialize (testRun) {
+        this.phase = PHASE.pendingInitialization;
+
+        await testRun.switchToCleanRun();
+
+        var navigateCommand = new NavigateToCommand({ url: this.loginPage });
+
+        await testRun.executeCommand(navigateCommand);
+
+        try {
+            await this.initFn(testRun);
+        }
+        catch (err) {
+            this.initErr = err;
+        }
+
+        if (!this.initErr) {
```

Comment:
Yes it can and will make code more compact, but it will ambiguous which code we expect will throw. It's a bad practice to put everything in the `try/catch`